Collectives framework: Reworked sched-cache

alexander-sannikov opened the pull request:
- cache implementation is transport specific
- reduced amount of global structures for better threading support (per-communicator cache)

Commit:
Collectives framework: Reworked sched-cache
- cache implementation is transport specific
- reduced amount of global structures for better threading support (per-communicator cache)

Commit:
Collectives framework: fixed memory deallocation issue.

akhillanger commented on src/mpi/coll/transports/mpich/transport.h:
```diff
@@ -170,17 +170,27 @@ static inline void MPIC_MPICH_free_buffers(MPIC_MPICH_sched_t * s)
+            MPIC_MPICH_free_mem(s->requests[i].nbargs.free_mem.ptr);
+            s->requests[i].nbargs.free_mem.ptr = NULL;
+        }
+    }
```
Please move this for loop inside MPIC_MPICH_free_buffers. There is already a for loop there, you can put this as an else if condition in that existing for loop

akhillanger commented on src/mpi/coll/algorithms/tree/coll.h:
```diff
@@ -51,13 +51,31 @@ static inline int COLL_allreduce(const void *sendbuf,
+    if (is_new) {
+        rc = COLL_sched_allreduce_tree(sendbuf, recvbuf, count,
+                    datatype, op, tag, comm, k, s, 1);
+        TSP_sched_cache_store(&comm->tsp_comm, (void*)&coll_args,
```
I want to rename this to TSP_save_schedule

akhillanger commented on src/mpi/coll/algorithms/tree/coll.h:
```diff
@@ -51,13 +51,31 @@ static inline int COLL_allreduce(const void *sendbuf,
-    COLL_sched_kick(&s);
+    /*Allocate or load from cache schedule (depends of transport implementation).
+     * Reset scheduler if necessary. */
+    TSP_sched_t *s = TSP_sched_get( &comm->tsp_comm, (void*) &coll_args,
```
Rename this to TSP_get_schedule

akhillanger commented on src/mpi/coll/algorithms/common/common_fns.h:
```diff
@@ -31,68 +31,55 @@
-    }
+    TSP_sched_start(s);
+    /* Enqueue nonblocking collective. */
+    request->phases = s;
```
We need to rename 'phases' to schedule. 'phases' is coming from Charles but now it does not make any sense.

akhillanger commented on src/mpi/coll/transports/mpich/transport.h:
```diff
@@ -82,15 +94,10 @@ static inline int MPIC_MPICH_init()
-static inline int MPIC_MPICH_comm_cleanup(MPIC_MPICH_comm_t * comm)
-{
-    comm->mpid_comm = NULL;
+    MPIC_add_sched( &(comm->sched_cache), key, len, s);
```
Since all this function is doing is making this function call, can we remove MPIC_add_sched and move its content  here.

alexander-sannikov replied:
MPIC_add_sched might be used by other transports, but with extra logic around - that's why it is separated.

akhillanger replied:
Ok.

akhillanger commented:
I am merging this PR, but @alexander-sannikov can you please address the comments and I will merge them as well.

Commit:
Collectives framework: review notes fixed.
TSP_sched_cache_store and TSP_sched_get are renamed.
MPIC_req_t type is generalized.

alexander-sannikov commented:
@akhillanger , PR is updated with one more commit.

akhillanger commented:
I have manually applied the last commit also now. Closing this PR.
